src/ext/php7/engine_hooks.c

```diff
@@ -545,5 +535,49 @@ void ddtrace_opcode_mshutdown(void) {
     zend_set_user_opcode_handler(ZEND_DO_FCALL, NULL);
     zend_set_user_opcode_handler(ZEND_DO_FCALL_BY_NAME, NULL);
 
+    zend_set_user_opcode_handler(ZEND_RETURN, NULL);
     zend_set_user_opcode_handler(ZEND_EXIT, NULL);
 }
+
+void ddtrace_execute_internal_minit(void) {
+    _prev_execute_internal = !zend_execute_internal ? execute_internal : zend_execute_internal;
+    zend_execute_internal = _dd_execute_internal;
+}
+
+void ddtrace_execute_internal_mshutdown(void) {
+    zend_execute_internal = _prev_execute_internal != execute_internal ? _prev_execute_internal : NULL;
+}
+
+static void _dd_execute_internal(zend_execute_data *execute_data, zval *return_value) {
+    zend_function *current_fbc = EX(func);
+    if (!current_fbc) {
+        _prev_execute_internal(execute_data, return_value);
+        return;
+    }
+    ddtrace_dispatch_t *dispatch = NULL;
+    if (!_dd_should_trace_call(execute_data, current_fbc, &dispatch)) {
+        _prev_execute_internal(execute_data, return_value);
+        return;
+    }
```

Comment:
What do you think about pushing the null check on the fbc into `_dd_should_trace_call`? This simplifies the caller and makes the call more robust, and I think this is the only caller of this function anyway. What do you think?

Reply:
That's where it used to be, but I had to move it out because in the function call hander, FBC lives at `EX(call)->func` and in the internal function hook it lives at `EX(func)`.